### ISSUE TITLE
API proposal : pooled / leased connections as a secondary API

### ISSUE THREAD

**Issue by mgravell** [StackExchange/StackExchange.Redis#886]
Motivations:

- provide safe access to blocking operations (`BRPOPLPUSH`, etc)
- provide "pure" access to `WATCH`/`MULTI`/`EXEC`

If implemented (needs consideration), the multiplexer would take care of the socket IO, but they would be stored separately and would not participate in the multiplexed command stream.

Questions:

- what does this mean for cluster? is it assumed that all operations on a lease will participate on a single shard and will therefore be single-server?
- how is primary/replica selected
- what does the lease API look like - presumably `using`?
- what happens when the pool is exhausted? is it an "await (async) an item from the pool, or timeout after x"?

**Comment by mgravell** [comment]
```
IServer server = conn.GetServer(key, flags); // args for sharding and primary/replica selection
using(ILeasedDatabase db = await server.Lease(timout)) // other args here not shown: db number, async state
{

}
```

with:

```
interface ILeasedDatabase : IDatabase, IDisposable
{
    // TODO: add blocking APIs

    // TODO: add "pure" WATCH/EXEC APIs
}
```

**Comment by mgravell** [comment]
@NickCraver  the motivation isn't data volume / load related; it is semantic related. There are a good few operations that are designed to be connection-blocking, so we don't expose them today. Similarly, there are some `WATCH` scenarios that we can't implement, because they require ad-hoc commands *inside* a `WATCH` region (but before a `MULTI`), which we disallow.

**Comment by aensidhe** [comment]
1. IDatabase is compatible with IDatabaseAsync. I can shoot myself into leg passing ILeasedDatabase into usual async code.
2. Why do we need separate API for creating additional connection? Why creating a separate multiplexer and caching it is worse? Chances are that I'll need it several times during lifetime of my web code. If code is in some tool, I can create as many multiplexers as I want.

My opinion that this API is unnecessary.

**Comment by NickCraver** [comment]
@mgravell I'm not sure I understand the intent of *where* you're picturing the leasing/blocking. It's a close in concept to a transaction (IMO), or the way I read it) we'd be blocking every command on the pipe behind that `using`, effectively making it `lock()` semantics w.r.t. the pipeline. Can you elaborate on the examples a bit more on where you expect the actual leasing/blocking to start and end? If it's not on a separate connection (pooled, dedicated, whatever), which most of my concerns are around, we're still back to blocking the existing primary connection (the primary reason, AFAIK, we disallow these today). Am I way off here?

**Comment by mgravell** [comment]
@aensidhe 

1. doesn't shoot you in the leg - it just means it runs on a separate connection
2. because establishing connections and doing all the handshaking is more expensive than you might think; and also: *we don't currently expose those APIs for you to use*

**Comment by mgravell** [comment]
@NickCraver  the leased connection *is a dedicated socket* - it is a leased *redis connection*, essentially; "interactive" only (no subscription functionality)

**Comment by NickCraver** [comment]
Okay so if I do this:
```c#
db.StringSet("mykey", "value");
using(ILeasedDatabase db = await server.LeaseAsync(timeout)) 
{
  var val = await db.StringGet("mykey");
}
```
...then `val` can be `null`, because we're got 2 connections and physical pipelines - so if the first is backed up a bit, it won't have issued the write to redis before the `GET` on the second, right? This is *less likely* to happen if using the shared `SocketManager` writer pools, but still. I think there's a lot of non-intuitive behaviors that could arise here: races, exhaustion of the pool (in loops/concurrency) or unexpected timeouts (assuming that's a concept that applies here), and other related bits when using these with any decent frequency.

Thoughts on those kinds of things?

**Comment by aensidhe** [comment]
@mgravell 

1. Yeah and that can be unexpected. Can't think about anything bad right now, so ok, let it be ok.
2. I know about cost of establishing of connection, but why is pooling connection is better than thinking ahead and putting an additional multiplexer ahead of time into DI container or in separate global variable, etc? This is very edge use case. Will you recycle pooled connection? How and when will you do it?

It's not a "user way of think", I'm writing a similar package for another database. And I think that if user want a separate connection from time to time he can cache that separate connection somewhere near main connection.

And other concern: `IDatabase` and `IDatabaseAsync` are lightweight wrapper to get and throw away here and there, if I'm remember it correctly. Making a compatible interface that can create a connection per allocation of that interface can be a problem.

**Comment by mgravell** [comment]
> Okay so if I do this:

No, there is no conflict there; `StringSet` blocks until the response comes *back* from the server; likewise `await StringSetAsync`. There *is* a thread-race if you do:

    var x = dbA.StringSetAsync();
    var y = dbB.StringGetAsync();
    await x;
    var s = await y;

but.... at that point you've gone out of your way to introduce a problem

**Comment by NickCraver** [comment]
Ah sorry I should amend the example (was on mobile), I don't think the `FireAndForget` is that uncommon/going out of the way - I think that's fairly standard fare on a set operations.

I'm just not (yet) convinced we've gotten enough feedback wanting this to justify spending much time on it, but I don't feel strongly against either. FWIW, some of the feedback like `WAIT` was around wanting things *on that pipeline* to finish (which is a separate issue from this anyway). I can recall a few issues related here, but IIRC they were all isolated.

My main concern is still around any sort of automatic *socket* pooling which makes things less predictable. That would feel (at this particular moment) like a step backwards. Right now failures there are failures on that multiplexer and it's not too hard to reason about what went wrong. When 1 (or many) in a pool of several start having issues, things get messy to figure out...that's the part that scares me, given we just got sockets decently stable and are still working on hangs (which to be fair may be unrelated). 

To be fair, it's entirely possible after we're stable on v2 I'd feel a lot more confident in diving into this with a different perspective on the complexity/debug risk vs. payoff.

**Comment by mgravell** [comment]
Fair enough; totally agree that it isn't a priority today - happy to shelve indefinitely, I just want to open the dialogue.

**Comment by Plasma** [comment]
Just wanted to chime in as to why having a separate connection, even for non-blocking operation use, can make sense some times.

At the moment the recommendation is to use a single, shared ConnectionMultiplexer across all threads.

This usually works well, but I've observed these side effects in production:

1) Because all commands from all threads go to a single C# queue and TCP socket, you get "head of line blocking" behind a buffer that may be filled with more commands from Thread A (perhaps megabytes in terms of bandwidth to process) in front of the one you're wanting to transmit that is more time sensitive, and smaller in size, from Thread B.

Imagine for example, Thread A has written several megabytes worth of SET commands due to a background task, and Thread B has a simple GET command pending at the end of the shared muxer queue. Thread B must wait for Thread A's workload to be emptied first.

The delays I am referring to here are milliseconds, and in one case 100s of milliseconds, due to CPU heavy redis commands used (eg sorted set intersects), but I'm mostly pointing it out in terms of "jitter" observed for random requests.

2) I acknowledge that Redis is single threaded, and processes requests sequentially, but it is worth noting that Redis attempts to treat each client/TCP connection fairly [1].

3) Imagine we utilized two connections ("clients") to Redis instead, and Thread A is doing its bandwidth-heavy, batch based processing on one connection, and Thread B uses another connection that is intended to be for more real-time request/response patterns.

If both are sent at the same moment, and my assumption is the local app server's TCP framework will also do interleaving, then Redis will switch/interleave processing requests from the TCP sockets between both connections, so while partially receiving/processing Thread A's workload (the multi-megabyte SET commands), Redis will process Thread B's simple workload quickly, then switch back to continuing to process Thread A workload.

The end result here is that Thread B gets a faster response time as its connection's workload was interleaved, on Redis's side, because it was a separate TCP connection, and it wasn't behind the workload of Thread A by sharing the same TCP socket or local C# command queue.

[1] https://redis.io/topics/clients

**Comment by mgravell** [comment]
Indeed. This is a scenario where we currently just spin up a second
parallel multiplexer - one for large slow ops, one for fast traffic. Just
mentioning that because it is a simple pragmatic low effort solution to the
same problems.

**Comment by prat0088** [comment]
Is this still on the table?  Is there anything I can help out with?

**Comment by mgravell** [comment]
"yes" and "probably not yet, later: definitely", in that order.

**Comment by DeepWorksStudios** [comment]
Any updates?

**Comment by mgravell** [comment]
Now that this is finally my day job, I hope we can improve that soon. I appreciate your frustration, but also keep in mind: except for the last few weeks, most of the effort here has come from scavenged personal time of a *very small* number of contributors, and the list of things to do has always been larger than thd available time.

**Comment by mgravell** [comment]
@aradalvand I have your question from earlier captured in my email. Nick wasn't wrong to shut it down. However: I'm "out" until Tuesday - I'll try to post a more... *constructive* version of it as a discussion when I'm back. Here's the short response I couldn't post because 404:

> I'm on the road at the moment - I will get back to you on Tuesday with *way* more of a response than is warranted, but the short version for now: you have just about every point backwards. I don't mean that in a rude or dismissive way - simply giving a heads-up of the direction my longer-form response will be taking.

I'll explain more then. More generally: maybe ask more questions rather than assuming you're right about everything.

I **will not** be discussing it more here. Please don't try.

**Comment by mgravell** [comment]
as promised: @aradalvand - https://github.com/StackExchange/StackExchange.Redis/discussions/3207; please feel free to join in there, but keep it constructive and civil
